source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+
+    Contains the quorum generator algorithm.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.Quorum;
+
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Set;
+import agora.common.Types;
+import agora.consensus.data.Enrollment;
+import agora.consensus.UTXOSet;
+
+import std.algorithm;
+import std.array;
+import std.exception;
+import std.format;
+import std.math;
+import std.range;
+import std.string;
+
+/*******************************************************************************
+
+    Build the quorum configuration for the given public key and the staked
+    enrollments.
+
+    Params:
+        node_key = the key of the node for which to generate the quorum
+        enrolls = the array of registered enrollments
+        finder = UTXO finder delegate
+
+    Returns:
+        the generated quorum config
+
+*******************************************************************************/
+
+public QuorumConfig buildQuorumConfig ( const ref PublicKey node_key,
+    in Enrollment[] enrolls, in UTXOFinder finder )
+{
+    Set!PublicKey all_nodes = getPublicKeys(enrolls, finder);
+
+    QuorumConfig quorum;
+    all_nodes.each!(node => quorum.nodes ~= node);
+    quorum.nodes.sort();
+    quorum.threshold = quorum.nodes.length;
+
+    return quorum;
+}
+
+///
+unittest
+{
+    // test up to 1024 nodes
+    foreach (num_nodes; iota(1, 10).map!(n => 2 ^^ n))
```

Comment:
// test up to 1022 nodes

```
unittest
{
    import std.algorithm;
    import std.stdio;
    import std.range;

    uint sum;
    foreach (num_nodes; iota(1, 10).map!(n => 2 ^^ n))
       sum+=num_nodes;
    writeln(sum);
}
```
Result  : 1022

Reply:
It's not testing all node count configurations. That would be too expensive (especially once the algorithm is changed).

What I am testing here is this:

```
num_nodes : 2
num_nodes : 4
num_nodes : 8
num_nodes : 16
num_nodes : 32
num_nodes : 64
num_nodes : 128
num_nodes : 256
num_nodes : 512
num_nodes : 1024
```

See?

Reply:
Well actually it's not inclusive as I forgot to use `11` in the iota call. +1 points to you.